gpii/node_modules/preferencesServer/test/prefsSafesTests.js

```diff
@@ -0,0 +1,715 @@
+            listener: "gpii.tests.preferencesServer.testResponse",
+            args: ["{arguments}.0", fixture.expected, "{getRequest}.nativeResponse.statusCode", expectedStatusCode]
+        }]
+    };
+};
+
+gpii.tests.preferencesServer.prefsSafes.getSafeWithKeys.successFixtures = [
+    {
+        name: "GET: Retieve a payload containing the prefs-safes and all it's associated key/credential records",
+        prefsSafeId: "prefsSafe-7",
+        url: "%prefsSafeId",
+        expected: {
+            prefsSafe: {
+                "id": "prefsSafe-7",
+                "type": "prefsSafe",
+                "schemaVersion": "0.1",
+                "prefsSafeType": "user",
+                "name": null,
+                "email": null,
+                "preferences": {
+                    "flat": {
+                        "name": "bit of stuff",
+                        "contexts": {
+                            "gpii-default": {
+                                "name": "Default preferences",
+                                "preferences": {
+                                    "http://registry.gpii.net/common/onScreenKeyboard/enabled": true
+                                },
+                                "metadata": []
+                            }
+                        },
+                        "metadata": []
+                    }
+                },
+                "timestampCreated": "2017-12-14T19:55:11.641Z",
+                "timestampUpdated": null
+            },
+            keys: [
+                {
+                    "type": "gpiiCloudSafeCredential",
+                    "schemaVersion": "0.1",
+                    "prefsSafeId": "prefsSafe-7",
+                    "gpiiExpressUserId": "org.couch.db.user:prefs7user",
+                    "id": "8f3085a7-b65b-4648-9a78-8ac7de766997"
+                },
+                {
+                    "id": "np_tiny",
+                    "type": "gpiiKey",
+                    "schemaVersion": "0.1",
+                    "prefsSafeId": "prefsSafe-7",
+                    "prefsSetId": "gpii-default",
+                    "revoked": false,
+                    "revokedReason": null,
+                    "timestampCreated": "2017-12-14T19:55:11.641Z",
+                    "timestampUpdated": null
+                }
+            ]
+        }
+    }
+];
+
+gpii.tests.preferencesServer.prefsSafes.getSafeWithKeys.failFixtures = [
+    {
+        name: "GET: Failure message for a prefs safe that does not exist.",
+        prefsSafeId: "prefsSafe-3999",
+        url: "%prefsSafeId",
+        expectedStatusCode: 404,
+        expected: {
+            "isError": true,
+            "message": "Cannot lookup prefsSafe"
+        }
+    }
+];
+
+/////////////////////// TESTING prefsSafes PUT (Save) ////////////////////////////
+fluid.registerNamespace("gpii.tests.preferencesServer.prefsSafes.putSafe");
+
+gpii.tests.preferencesServer.prefsSafes.putSafe.buildTestDef = function (fixture) {
```

Comment:
A couple of issues with this endpoint. Please also write tests to cover these cases:

1. Sending this request with an non-existent prefs safe id will have the system auto create a new prefs safe with the given id. Is this what you want? The documentation says this endpoint is for updating existing safes. Please fix either the code or the documentation.
2. The code doesn't check if the prefsSafeId sent along as a request parameter matches the prefsSafeId in the PUT body. If there's a mismatch, the system still goes ahead to update the database with the PUT body.

Reply:
I've updated the docs that this endpoint is _only_ for updating, and have added checks and unit tests for both of the above scenerios.